pkg/deprecation/deprecation.go

```diff
@@ -0,0 +1,365 @@
+	message := "DEPRECATION WARNING"
+	if messageHeader != "" {
+		message += ": " + messageHeader
+	}
+	message += sunsetMessage
+
+	fmt.Fprintf(os.Stderr, "%s\n", message)
+}
+
+// GlobalWarningHandler is the global handler for deprecation warnings
+var GlobalWarningHandler WarningHandler = &DefaultWarningHandler{}
+
+// SetWarningHandler allows users to set a custom deprecation warning handler
+func SetWarningHandler(handler WarningHandler) {
+	GlobalWarningHandler = handler
+}
+
+// ResponseWithHeaders represents any response type that has headers
+type ResponseWithHeaders interface {
+	Header() http.Header
+}
+
+// RequestInterface represents any request that can be sent
+type RequestInterface interface {
+	Send() (*sdk.Response, error)
+}
+
+// TypedRequestInterface represents typed requests (like Get, Post, etc.)
+type TypedRequestInterface[T any] interface {
+	Send() (T, error)
+}
+
+// HandleFromHeaders checks for deprecation headers and handles them appropriately
+func HandleFromHeaders(headers http.Header) {
+	if GlobalWarningHandler == nil {
+		return
+	}
+
+	var deprecationHeader, messageHeader string
+
+	// Look for deprecation headers - use direct access since Get() seems to have issues
+	if values := headers[consts.DeprecationHeader]; len(values) > 0 {
+		deprecationHeader = values[0]
+	}
+	if values := headers[consts.OCMDeprecationMessage]; len(values) > 0 {
+		messageHeader = values[0]
+	}
+
+	GlobalWarningHandler.HandleWarning(deprecationHeader, messageHeader)
+}
+
+// HandleDeprecationWarning checks for deprecation headers in the HTTP response
+// and prints appropriate warning messages.
+func HandleDeprecationWarning(response *sdk.Response) {
+	if response == nil {
+		return
+	}
+
+	getHeader := response.Header
+	if getHeader == nil {
+		return
+	}
+
+	// Look for deprecation headers
+	deprecationHeader := getHeader(consts.DeprecationHeader)
+	messageHeader := getHeader(consts.OCMDeprecationMessage)
+
+	if GlobalWarningHandler != nil {
+		GlobalWarningHandler.HandleWarning(deprecationHeader, messageHeader)
+	}
+}
+
+// HandleFromResponse handles deprecation warnings from SDK responses
+func HandleFromResponse(response *sdk.Response) {
+	if response == nil {
+		return
+	}
+
+	getHeader := response.Header
+	if getHeader == nil {
+		return
+	}
+
+	// Look for deprecation headers
+	deprecationHeader := getHeader(consts.DeprecationHeader)
+	messageHeader := getHeader(consts.OCMDeprecationMessage)
+
+	if GlobalWarningHandler != nil {
+		GlobalWarningHandler.HandleWarning(deprecationHeader, messageHeader)
+	}
+}
+
+// HandleDeprecationWarningFromTypedResponse handles deprecation warnings from typed responses
+func HandleDeprecationWarningFromTypedResponse(response ResponseWithHeaders) {
+	if response == nil {
+		return
+	}
+
+	headers := response.Header()
+	HandleFromHeaders(headers)
+}
+
+// HandleFromTypedResponse handles deprecation warnings from typed responses
+func HandleFromTypedResponse(response ResponseWithHeaders) {
```

Comment:
Can we decompose these a little, the logic here is the same as `HandleDeprecationWarningFromTypedResponse`